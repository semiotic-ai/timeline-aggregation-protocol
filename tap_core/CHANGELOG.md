# Changelog

## [2.0.0](https://github.com/semiotic-ai/timeline-aggregation-protocol/compare/tap_core-v1.0.0...tap_core-v2.0.0) (2024-10-30)


### ⚠ BREAKING CHANGES

* add context to checks
* add retryable errors to checks

### Features

* add context to checks ([58a6a52](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/58a6a52eba8152ee3add27d10e59b9e10ab9a5f4))
* add retryable errors to checks ([51f04cb](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/51f04cb0dbe7387ffd94f16eb59bbeb1c6c51680))
* expected_rav to become result&lt;rav,error&gt; ([70fa8fa](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/70fa8fa116c32b074f5c691979de02177b385560))
* make error public access on failed receipt ([7180b86](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/7180b86e2c5ef5e2009518d6cd2ed8fe16b337eb))
* obtain invalid receipts ([101dc5e](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/101dc5eea45baf143c5135fe66bbb5f62dfa9109))


### Bug Fixes

* Remove obtain_invalid_receipts ([ad255dc](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/ad255dcbb4a5ea101c575038534312b59e12d693))

## [1.0.0](https://github.com/semiotic-ai/timeline-aggregation-protocol/compare/tap_core-v0.8.0...tap_core-v1.0.0) (2024-03-27)


### ⚠ BREAKING CHANGES

* create docs and readme.

### Documentation

* create docs and readme. ([4ebf258](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/4ebf2584441514166fd9e91999b1710af1d54524))

## [0.8.0](https://github.com/semiotic-ai/timeline-aggregation-protocol/compare/tap_core-v0.7.0...tap_core-v0.8.0) (2024-03-11)


### ⚠ BREAKING CHANGES

* update project structure
* implement unique and timestamp check hard coded
* remove stateful checks
* remove timestamp check from manager
* update rav attributes to camel case
* use single executor for manager and auditor
* convert receipt and rav storage into executor
* use typestate for receivedreceipt
* split read and write storage adapters

### Features

* add serde to ReceivedReceipt ([b13bedf](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/b13bedf4f145fe0209c48d4d51630233ad71b4b8))
* **aggregator:** Add support for multiple signers in input ([#211](https://github.com/semiotic-ai/timeline-aggregation-protocol/issues/211)) ([b16f23d](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/b16f23d5481de65658b08544c083e2849821370e)), closes [#205](https://github.com/semiotic-ai/timeline-aggregation-protocol/issues/205)
* **core:** Supply EIP712 domain sep with prefilled version and name ([#210](https://github.com/semiotic-ai/timeline-aggregation-protocol/issues/210)) ([2ed564b](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/2ed564b9581c4eff2364fdc490f0f8d0022a6982))
* implement unique and timestamp check hard coded ([7c3f5a9](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/7c3f5a97679b7fcb3cd1150517e6bd77585255e0))


### Bug Fixes

* add debug to messageid ([fbdd328](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/fbdd328f4170d58800b62deed833493b728afb68))
* checks need to be send and sync ([78ae2a8](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/78ae2a89e395a49a186eef62a26c5f51ca279735))
* compile tests ([06cf24c](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/06cf24cac1b4e50bcd8234557e09556b7de75007))
* implement receipt delete to mock ([50058f7](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/50058f7043303bbb914de7808b641b39800e0ede))
* missing merge conflicts ([8c95954](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/8c9595456bad9c334b539d0363de884239f14157))
* rebase main ([0c7bc1c](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/0c7bc1c9cebbb8dfad1c83067a1457380009925f))
* remove EscrowAdapter dependency ([3045f61](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/3045f61c2ce556ab13bf9268102c9af0387b6226))
* update rav attributes to camel case ([1b232e4](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/1b232e4de230dc4922937fea2b489c7409ca2408))
* update timestamp check only after rav update ([bfb1eb0](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/bfb1eb07aa2d26f611ce14d4d0b57fddb9da2ea6))
* use the correct timestamp check ([1b1d19c](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/1b1d19ca00244f01fb45c8d62398dfe16ddac7bb))


### Reverts

* rename back to adapters ([42bd2df](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/42bd2dfce16d12753ade4577cb0ee47b3d82dbce))


### Code Refactoring

* convert receipt and rav storage into executor ([467c917](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/467c917af61733e3c2cbf3823c4377091179980b))
* remove stateful checks ([1044f53](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/1044f534be5e35a387d32e6b8bd86e86af188b66))
* remove timestamp check from manager ([0a3b983](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/0a3b983b5f20b27c0afdc02d9db682481e7493a9))
* split read and write storage adapters ([2e681a5](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/2e681a529d0c03e2fe455f64692e23024ad98073))
* update project structure ([70ee2c6](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/70ee2c67125653c25f479ee5f11e7c7e555078b7))
* use single executor for manager and auditor ([6794fbb](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/6794fbb6a7d02774008d57d62ccc10564cf952c1))
* use typestate for receivedreceipt ([89b5d94](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/89b5d941a37c475cb47c768f2618b902cf8908c5))

## [0.7.0](https://github.com/semiotic-ai/timeline-aggregation-protocol/compare/tap_core-v0.6.0...tap_core-v0.7.0) (2023-11-28)


### ⚠ BREAKING CHANGES

* add limit to receipts retrieve
* create_rav_request() returns invalid ReceivedReceipt
* rename "gateway" to "sender" everywhere

### Features

* add limit to receipts retrieve ([0ce2aab](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/0ce2aabca5cf20bb531578e3a92d33ed0a2d4c17))
* create_rav_request() returns invalid ReceivedReceipt ([5bb9001](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/5bb90015dc636aa4c0916e7072724ae7fc03e8e4))


### Bug Fixes

* mutable manager.remove_obsolete_receipts ([ca1a01e](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/ca1a01e8571f9c65a6b58efbb4964b568c03e6ac))


### Code Refactoring

* rename "gateway" to "sender" everywhere ([309f41f](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/309f41f879b51a1f2840ef0ed2552d7faa338b86)), closes [#188](https://github.com/semiotic-ai/timeline-aggregation-protocol/issues/188)

## [0.6.0](https://github.com/semiotic-ai/timeline-aggregation-protocol/compare/tap_core-v0.5.1...tap_core-v0.6.0) (2023-10-12)


### ⚠ BREAKING CHANGES

* **core:** 

### Code Refactoring

* **core:** manager arg slice instead of vec ([ee95e94](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/ee95e9492bdfeeabadc2b0510940abc082b34cd4))

## [0.5.1](https://github.com/semiotic-ai/timeline-aggregation-protocol/compare/tap_core-v0.5.0...tap_core-v0.5.1) (2023-08-24)


### Bug Fixes

* Update alloy to v0.3.2 ([715a04c](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/715a04c1d9b9622e0bb846830f8dd782062901c7))

## [0.5.0](https://github.com/semiotic-ai/timeline-aggregation-protocol/compare/tap_core-v0.4.0...tap_core-v0.5.0) (2023-08-17)


### ⚠ BREAKING CHANGES

* **tap-core:** receipts, RAVs, and eip712-signed-message now use alloy primitive address

### Features

* **tap-core-bench:** update to work with alloy changes ([43c5e6c](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/43c5e6c24218b9c5138cabec50262715f6b34124))
* **tap-core:** changes eip712 and address components to alloy library ([a99af2f](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/a99af2f8d64ad2d1a4d175de5e7789254fb2c36f))

## [0.4.0](https://github.com/semiotic-ai/timeline-aggregation-protocol/compare/tap_core-v0.3.0...tap_core-v0.4.0) (2023-08-08)


### ⚠ BREAKING CHANGES

* **core:** All ReceiptChecksAdapter trait method now return Result. Makes it possible to return Adapter errors besides just check failures.
* All instances of the word "collateral" are replaced with "escrow".

### Code Refactoring

* **core:** make ChecksAdapter return Result ([0983c27](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/0983c27c40d2061a51687fa0517add2593d3b934))
* Replace "collateral" with "escrow" ([6f9d0c7](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/6f9d0c7a54fa445f1e058a65bcfccdef06cf9d02))

## [0.3.0](https://github.com/semiotic-ai/timeline-aggregation-protocol/compare/tap_core-v0.2.0...tap_core-v0.3.0) (2023-07-31)


### ⚠ BREAKING CHANGES

* **core:** all the traits are now async

### Features

* serde for ReceivedReceipt ([1765b5d](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/1765b5d101e73ff0085e6246252b47e4d5e98890))


### Code Refactoring

* **core:** make it all async ([30ca4ba](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/30ca4ba7da05959d380b653de0307af273c4733d))

## [0.2.0](https://github.com/semiotic-ai/timeline-aggregation-protocol/compare/tap_core-v0.1.0...tap_core-v0.2.0) (2023-07-20)


### ⚠ BREAKING CHANGES

* **tap_core:** Preserve adapter error type using anyhow ([#133](https://github.com/semiotic-ai/timeline-aggregation-protocol/issues/133))
* **receipt_storage_adapter:** all functions unnecessary to the TAP manager have been removed from the `ReceiptStorageAdapter` trait.
* **rav_storage_adapter:** the RavStorageAdapter public trait has changed

### Features

* **manager:** add receipts_auto_delete ([#130](https://github.com/semiotic-ai/timeline-aggregation-protocol/issues/130)) ([37bc8e1](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/37bc8e1212fb525e15be465bdc0dba4e5349c025))
* **receipt_storage_adapter:** Use RangeBounds ([663a8ba](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/663a8ba7f0afb74cc17e4eb58802282771cb0bca))
* **tap_core:** Preserve adapter error type using anyhow ([#133](https://github.com/semiotic-ai/timeline-aggregation-protocol/issues/133)) ([77abbd8](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/77abbd82d5e0c75cd04b1bce55360a312f036bac))


### Bug Fixes

* **tap_manager:** receipt auditor min_timestamp set incorrectly ([743ac7c](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/743ac7cd252a3723ad01fbd4a1b54cbaf68267af))


### Code Refactoring

* **rav_storage_adapter:** simplify trait ([6af9471](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/6af9471bea8aeafd415b4f3e7b92a1eb6bec135b))
* **receipt_storage_adapter:** prune trait ([caef197](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/caef197261f16187402f69091ffea843e6aef970))

## 0.1.0 (2023-06-29)


### ⚠ BREAKING CHANGES

* **tap-manager:** RAV Request definition is updated to include optional previous RAV
* **receipts:** Updates the receipt checks adapter api
* **signed-message:** Updates signed message API
* **allocation-adapter:** removes allocation adapter
* **adapters:** existing adapter trait definitions are updated

### Features

* **adapter-mocks:** alters adapters to use references (arc) to allow sharing common resources ([03e7668](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/03e7668e3c59d27e6cfc869a3a35ad1434d18d6d))
* **adapters:** adds functionality to retrieve and delete receipts within a range ([4143ac6](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/4143ac6293751a0e837709bba43d4fc600911bcc))
* **adapters:** split adapters into storage and check adapters ([39c1c82](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/39c1c827447aceb938ad03643bb7bf08ff330cae))
* **core:** add `verify` to EIP712 signed msg ([a7e3e7d](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/a7e3e7d18044dbe6937cf725376167171fb177b1))
* **receipt-storage:** Adds functionality to update receipt in storage by id ([eb4f8ba](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/eb4f8bae233406b6c5d25def4de1d628d7860b1e))
* **receipts:** Updates checking mechanisms and adds auditor to complete checks ([8d0088d](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/8d0088d6fbc83416737cf33c3e305412741c8ec8))
* **signed-message:** Updates library to use ether-rs for wallet, address, key, signature, and verification ([7f1cb85](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/7f1cb8586e7577221008588cacd5cc6ad47d1c83))
* **tap-manager:** Adds a tap manager for handling receipt and rav validation and storage ([3786042](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/378604263a91d3abd9fdad1dd978f1ba715f7aca))


### Bug Fixes

* **allocation-adapter:** remove obsolete trait ([957f3f9](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/957f3f99efa8b5ebe7f61024f96905b0448c4bed))
* **receipt-errors:** updates receipt errors to work with adapter trait api ([fc121bf](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/fc121bf21bee2b8bdf0d1db026e97afb2844d75b))
* **receipts:** adds receipt ID to check unique to allow checking after receipt is in storage ([5072fb9](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/5072fb9ba614d58bcc712778deb451eaefbc993f))
* **tap-manager:** adds an error when receipt timestamp window is inverted ([e21d7d9](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/e21d7d941b8b298bdda5dc8143b3163f65ca1e85))
* **tap-manager:** receipts being used after being added to RAV ([efd88a2](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/efd88a214a3737b7bb201cabaf4037284ec5d547))
* verification benchmarks ([#114](https://github.com/semiotic-ai/timeline-aggregation-protocol/issues/114)) ([96cdf24](https://github.com/semiotic-ai/timeline-aggregation-protocol/commit/96cdf24db98a715cec654ff77de1837ba36f81a4))
