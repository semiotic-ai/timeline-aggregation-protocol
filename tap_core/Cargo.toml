[package]
name = "tap_core"
version = "2.0.0"
edition.workspace = true
license.workspace = true
repository.workspace = true
readme = "README.md"
description = "Core Timeline Aggregation Protocol library: a fast, efficient and trustless unidirectional micro-payments system."

[dependencies]
alloy.workspace = true
serde.workspace = true
tokio.workspace = true
anyhow.workspace = true
rand.workspace = true
thiserror = "1.0.38"
async-trait = "0.1.72"
anymap3 = "1.0.0"

[dev-dependencies]
criterion = { version = "0.5", features = ["async_std"] }
rstest.workspace = true


[features]
default = ["in_memory"]
in_memory = []

[[bench]]
name = 'timeline_aggretion_protocol_benchmark'
harness = false
