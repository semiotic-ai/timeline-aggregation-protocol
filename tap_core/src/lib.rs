// Copyright 2023-, Semiotic AI, Inc.
// SPDX-License-Identifier: Apache-2.0

//! The Timeline Aggregation Protocol (TAP) is a micro-trust
//! state channel payment solution allowing one-way payments
//! from a payment sender to be aggregated then cheaply
//! verified on-chain by a payment receiver.

use std::time::{SystemTime, UNIX_EPOCH};

use thiserror::Error;

pub mod adapters;
pub mod eip_712_signed_message;
mod error;
pub mod receipt_aggregate_voucher;
pub mod tap_manager;
pub mod tap_receipt;

pub use error::{Error, Result};

pub(crate) fn get_current_timestamp_u64_ns() -> Result<u64> {
    Ok(SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_err(|err| Error::InvalidSystemTime {
            source_error_message: err.to_string(),
        })?
        .as_nanos() as u64)
}

#[cfg(test)]
mod tap_tests {
    use std::str::FromStr;

    use ethereum_types::Address;
    use ethers::signers::{coins_bip39::English, LocalWallet, MnemonicBuilder, Signer};
    use rstest::*;

    use crate::{
        eip_712_signed_message::EIP712SignedMessage,
        receipt_aggregate_voucher::ReceiptAggregateVoucher, tap_receipt::Receipt,
    };

    #[fixture]
    fn keys() -> (LocalWallet, Address) {
        let wallet: LocalWallet = MnemonicBuilder::<English>::default()
         .phrase("abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about")
         .build()
         .unwrap();
        let address = wallet.address();
        (wallet, address)
    }

    #[fixture]
    fn allocation_ids() -> Vec<Address> {
        vec![
            Address::from_str("0xabababababababababababababababababababab").unwrap(),
            Address::from_str("0xdeaddeaddeaddeaddeaddeaddeaddeaddeaddead").unwrap(),
            Address::from_str("0xbeefbeefbeefbeefbeefbeefbeefbeefbeefbeef").unwrap(),
            Address::from_str("0x1234567890abcdef1234567890abcdef12345678").unwrap(),
        ]
    }

    #[rstest]
    #[case::basic_rav_test (vec![45,56,34,23])]
    #[case::rav_from_zero_valued_receipts (vec![0,0,0,0])]
    async fn signed_rav_is_valid_with_no_previous_rav(
        keys: (LocalWallet, Address),
        allocation_ids: Vec<Address>,
        #[case] values: Vec<u128>,
    ) {
        // Create receipts
        let mut receipts = Vec::new();
        for value in values {
            receipts.push(
                EIP712SignedMessage::new(Receipt::new(allocation_ids[0], value).unwrap(), &keys.0)
                    .await
                    .unwrap(),
            );
        }

        // Skipping receipts validation in this test, aggregate_receipts assumes receipts are valid.

        let rav = ReceiptAggregateVoucher::aggregate_receipts(allocation_ids[0], &receipts, None)
            .unwrap();
        let signed_rav = EIP712SignedMessage::new(rav, &keys.0).await.unwrap();
        assert!(signed_rav.recover_signer().unwrap() == keys.1);
    }

    #[rstest]
    #[case::basic_rav_test(vec![45,56,34,23])]
    #[case::rav_from_zero_valued_receipts(vec![0,0,0,0])]
    async fn signed_rav_is_valid_with_previous_rav(
        keys: (LocalWallet, Address),
        allocation_ids: Vec<Address>,
        #[case] values: Vec<u128>,
    ) {
        // Create receipts
        let mut receipts = Vec::new();
        for value in values {
            receipts.push(
                EIP712SignedMessage::new(Receipt::new(allocation_ids[0], value).unwrap(), &keys.0)
                    .await
                    .unwrap(),
            );
        }

        // Create previous RAV from first half of receipts
        let prev_rav = ReceiptAggregateVoucher::aggregate_receipts(
            allocation_ids[0],
            &receipts[0..receipts.len() / 2],
            None,
        )
        .unwrap();
        let signed_prev_rav = EIP712SignedMessage::new(prev_rav, &keys.0).await.unwrap();

        // Create new RAV from last half of receipts and prev_rav
        let rav = ReceiptAggregateVoucher::aggregate_receipts(
            allocation_ids[0],
            &receipts[receipts.len() / 2..receipts.len()],
            Some(signed_prev_rav),
        )
        .unwrap();
        let signed_rav = EIP712SignedMessage::new(rav, &keys.0).await.unwrap();

        assert!(signed_rav.recover_signer().unwrap() == keys.1);
    }

    #[rstest]
    async fn verify_signature(keys: (LocalWallet, Address), allocation_ids: Vec<Address>) {
        let signed_message =
            EIP712SignedMessage::new(Receipt::new(allocation_ids[0], 42).unwrap(), &keys.0)
                .await
                .unwrap();

        assert!(signed_message.verify(keys.1).is_ok());
        assert!(signed_message
            .verify(Address::from_str("0x76f4eeD9fE41262669D0250b2A97db79712aD855").unwrap())
            .is_err());
    }
}
