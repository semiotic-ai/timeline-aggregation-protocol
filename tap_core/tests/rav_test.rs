// Copyright 2023-, Semiotic AI, Inc.
// SPDX-License-Identifier: Apache-2.0

use std::{
    collections::HashMap,
    str::FromStr,
    sync::{Arc, RwLock},
};

#[allow(deprecated)]
use alloy::primitives::{Address, PrimitiveSignature, Signature};
use alloy::{
    dyn_abi::Eip712Domain,
    signers::local::{coins_bip39::English, MnemonicBuilder, PrivateKeySigner},
};
use rstest::*;
use tap_core::{
    manager::{
        adapters::{RAVRead, RAVStore},
        context::memory::InMemoryContext,
    },
    rav::ReceiptAggregateVoucher,
    receipt::{checks::StatefulTimestampCheck, Receipt},
    signed_message::EIP712SignedMessage,
    tap_eip712_domain,
};

#[fixture]
fn domain_separator() -> Eip712Domain {
    tap_eip712_domain(1, Address::from([0x11u8; 20]))
}

#[fixture]
fn context() -> InMemoryContext {
    let escrow_storage = Arc::new(RwLock::new(HashMap::new()));
    let rav_storage = Arc::new(RwLock::new(None));
    let receipt_storage = Arc::new(RwLock::new(HashMap::new()));

    let timestamp_check = Arc::new(StatefulTimestampCheck::new(0));
    InMemoryContext::new(
        rav_storage,
        receipt_storage.clone(),
        escrow_storage.clone(),
        timestamp_check,
    )
}

#[rstest]
fn check_for_rav_serialization(domain_separator: Eip712Domain) {
    let allocation_id = Address::from_str("0xabababababababababababababababababababab").unwrap();
    let wallet = MnemonicBuilder::<English>::default()
        .phrase("abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about")
        .build()
        .unwrap();
    let mut receipts = Vec::new();
    for value in 50..60 {
        receipts.push(
            EIP712SignedMessage::new(
                &domain_separator,
                Receipt {
                    allocation_id,
                    value,
                    nonce: value as u64,
                    timestamp_ns: value as u64,
                },
                &wallet,
            )
            .unwrap(),
        );
    }

    let signed_rav = EIP712SignedMessage::new(
        &domain_separator,
        ReceiptAggregateVoucher::aggregate_receipts(allocation_id, &receipts, None).unwrap(),
        &wallet,
    )
    .unwrap();

    insta::assert_json_snapshot!(receipts);
    insta::assert_json_snapshot!(signed_rav);

    let raw_sig = r#"{
      "r": "0x1596dd0d380ede7aa5dec5ed09ea7d1fa8e4bc8dfdb43a4e965bb4f16906e321",
      "s": "0x788b69625a031fbd2e769928b63505387df16e7c51f19ff67c782bfec101a387",
      "yParity": "0x1"
    }"#;

    serde_json::from_str::<PrimitiveSignature>(raw_sig).unwrap();
    #[allow(deprecated)]
    serde_json::from_str::<Signature>(raw_sig).unwrap();
}

#[rstest]
#[tokio::test]
async fn rav_storage_adapter_test(domain_separator: Eip712Domain, context: InMemoryContext) {
    let wallet = PrivateKeySigner::random();

    let allocation_id = Address::from_str("0xabababababababababababababababababababab").unwrap();

    // Create receipts
    let mut receipts = Vec::new();
    for value in 50..60 {
        receipts.push(
            EIP712SignedMessage::new(
                &domain_separator,
                Receipt::new(allocation_id, value).unwrap(),
                &wallet,
            )
            .unwrap(),
        );
    }

    let signed_rav = EIP712SignedMessage::new(
        &domain_separator,
        ReceiptAggregateVoucher::aggregate_receipts(allocation_id, &receipts, None).unwrap(),
        &wallet,
    )
    .unwrap();

    context.update_last_rav(signed_rav.clone()).await.unwrap();

    // Retreive rav
    let retrieved_rav = context.last_rav().await;
    assert!(retrieved_rav.unwrap().unwrap() == signed_rav);

    // Testing the last rav update...

    // Create more receipts
    let mut receipts = Vec::new();
    for value in 60..70 {
        receipts.push(
            EIP712SignedMessage::new(
                &domain_separator,
                Receipt::new(allocation_id, value).unwrap(),
                &wallet,
            )
            .unwrap(),
        );
    }

    let signed_rav = EIP712SignedMessage::new(
        &domain_separator,
        ReceiptAggregateVoucher::aggregate_receipts(allocation_id, &receipts, None).unwrap(),
        &wallet,
    )
    .unwrap();

    // Update the last rav
    context.update_last_rav(signed_rav.clone()).await.unwrap();

    // Retreive rav
    let retrieved_rav = context.last_rav().await;
    assert!(retrieved_rav.unwrap().unwrap() == signed_rav);
}
