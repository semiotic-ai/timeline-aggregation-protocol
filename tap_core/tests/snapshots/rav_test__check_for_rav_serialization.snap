---
source: tap_core/tests/rav_test.rs
expression: receipts
---
[
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "payer": "0x0000000000000000000000000000000000000000",
      "data_service": "0x0000000000000000000000000000000000000000",
      "service_provider": "0x0000000000000000000000000000000000000000",
      "timestamp_ns": 1000000050,
      "nonce": 50,
      "value": "0x32"
    },
    "signature": {
      "r": "0x598d7cafa9645ac65e9f8a6c3834f36baee13b887c341dd09b5f92615bd6a11d",
      "s": "0x32410495bbcd25ba972396b74a2a6148da6f8ed7a89a6ae21ae87313fcce3ffb",
      "yParity": "0x0",
      "v": "0x0"
    }
  },
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "payer": "0x0000000000000000000000000000000000000000",
      "data_service": "0x0000000000000000000000000000000000000000",
      "service_provider": "0x0000000000000000000000000000000000000000",
      "timestamp_ns": 1000000051,
      "nonce": 51,
      "value": "0x33"
    },
    "signature": {
      "r": "0x5d9e36393ff0a43e096af18bddded1a5194fd731f215a3deb2f88c401894f175",
      "s": "0x720ec0c96bfb9bf149c27df7b6a62b1634fc0aa942074a7b7b5cc7c149e53181",
      "yParity": "0x1",
      "v": "0x1"
    }
  },
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "payer": "0x0000000000000000000000000000000000000000",
      "data_service": "0x0000000000000000000000000000000000000000",
      "service_provider": "0x0000000000000000000000000000000000000000",
      "timestamp_ns": 1000000052,
      "nonce": 52,
      "value": "0x34"
    },
    "signature": {
      "r": "0x9dc0db0a9b27d0379f2f94d5ddd58e67525db6eff39f292c431a45d08cfa704e",
      "s": "0x183dd49a895219314283c20d850aa59b06539a1399d80e81d6b5c251bf3c9061",
      "yParity": "0x0",
      "v": "0x0"
    }
  },
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "payer": "0x0000000000000000000000000000000000000000",
      "data_service": "0x0000000000000000000000000000000000000000",
      "service_provider": "0x0000000000000000000000000000000000000000",
      "timestamp_ns": 1000000053,
      "nonce": 53,
      "value": "0x35"
    },
    "signature": {
      "r": "0xb813aaa57768feea8f7aed2112a10deece693def87418140e67eb0b9bd8024a7",
      "s": "0x6748e6144879fcd01608e66ad117d7fbd2cca0f47e14058a89dc61afc5a10acd",
      "yParity": "0x0",
      "v": "0x0"
    }
  },
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "payer": "0x0000000000000000000000000000000000000000",
      "data_service": "0x0000000000000000000000000000000000000000",
      "service_provider": "0x0000000000000000000000000000000000000000",
      "timestamp_ns": 1000000054,
      "nonce": 54,
      "value": "0x36"
    },
    "signature": {
      "r": "0xaa2edc4947e70dd58dc40870ad8af320fb8bc448fed4f3a661c3bde075b10db8",
      "s": "0x6d8ac6b91a3fbcefabcf87b6d07c49b8f69994243b548f51558389c935a2079e",
      "yParity": "0x1",
      "v": "0x1"
    }
  },
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "payer": "0x0000000000000000000000000000000000000000",
      "data_service": "0x0000000000000000000000000000000000000000",
      "service_provider": "0x0000000000000000000000000000000000000000",
      "timestamp_ns": 1000000055,
      "nonce": 55,
      "value": "0x37"
    },
    "signature": {
      "r": "0xc48d72476911826acadb88bf6e8042e32532ef44d80dfec4cd67cbb1252a9829",
      "s": "0x752fc77426fe4aa0eba0409d866ffa72fc9351ac1627ca4c0f67d626d4045e48",
      "yParity": "0x0",
      "v": "0x0"
    }
  },
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "payer": "0x0000000000000000000000000000000000000000",
      "data_service": "0x0000000000000000000000000000000000000000",
      "service_provider": "0x0000000000000000000000000000000000000000",
      "timestamp_ns": 1000000056,
      "nonce": 56,
      "value": "0x38"
    },
    "signature": {
      "r": "0xbe8e91c2fc53ed4decfae0a02649fc029463d5e38f16bab537c915f52e13ff77",
      "s": "0x3e65f3cdc5dea626307728430bbc11eba5529781ddb903ded691be1b0c75917e",
      "yParity": "0x1",
      "v": "0x1"
    }
  },
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "payer": "0x0000000000000000000000000000000000000000",
      "data_service": "0x0000000000000000000000000000000000000000",
      "service_provider": "0x0000000000000000000000000000000000000000",
      "timestamp_ns": 1000000057,
      "nonce": 57,
      "value": "0x39"
    },
    "signature": {
      "r": "0x1d3672ecb9721dc8787af90bb12d5f09d961d33ae89b8b15b9ee8a492a09fe1d",
      "s": "0x164828da427829888a430110a49fe20bbd13076959ae91d6c001732fe7d33c45",
      "yParity": "0x1",
      "v": "0x1"
    }
  },
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "payer": "0x0000000000000000000000000000000000000000",
      "data_service": "0x0000000000000000000000000000000000000000",
      "service_provider": "0x0000000000000000000000000000000000000000",
      "timestamp_ns": 1000000058,
      "nonce": 58,
      "value": "0x3a"
    },
    "signature": {
      "r": "0x533acf762d392711a0887370ac40da601d55a9e9fc3f4416d935ff69403cb31f",
      "s": "0x172b4bc377ea7337eb0af260f28704daef630c148161e1254491a43781bb2dba",
      "yParity": "0x1",
      "v": "0x1"
    }
  },
  {
    "message": {
      "allocation_id": "0xabababababababababababababababababababab",
      "payer": "0x0000000000000000000000000000000000000000",
      "data_service": "0x0000000000000000000000000000000000000000",
      "service_provider": "0x0000000000000000000000000000000000000000",
      "timestamp_ns": 1000000059,
      "nonce": 59,
      "value": "0x3b"
    },
    "signature": {
      "r": "0x1ab9bc3b61e2faef9d631a8ee765a591da784dbad5eed6672a4c369b5d25e178",
      "s": "0x1535c22e79fba81e19d0d69cf29dfcf5b6b2faf3fa5203a6285aecd6dfb9af03",
      "yParity": "0x0",
      "v": "0x0"
    }
  }
]
